internal/cgroups/cgroups2.go

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) 2022 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+//go:build linux
+// +build linux
+
+package cgroups
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"strconv"
+	"strings"
+)
+
+const (
+	// _cgroupv2CPUMax is the file name for the CGroup-V2 CPU max and period
+	// parameter.
+	_cgroupv2CPUMax = "cpu.max"
+	// _cgroupFSType is the Linux CGroup-V2 file system type used in
+	// `/proc/$PID/mountinfo`.
+	_cgroupv2FSType = "cgroup2"
+
+	_cgroupv2MountPoint = "/sys/fs/cgroup"
+
+	_cgroupV2CPUMaxDefaultPeriod = 100000
+	_cgroupV2CPUMaxQuotaMax      = "max"
+)
+
+const (
+	_cgroupv2CPUMaxQuotaIndex = iota
+	_cgroupv2CPUMaxPeriodIndex
+)
+
+// IsCGroupV2 returns true if the system supports and uses cgroup2.
+// It gets the required information for deciding from mountinfo file.
+func IsCGroupV2() (bool, error) {
+	return isCGroupV2(_procPathMountInfo)
+}
+
+func isCGroupV2(procPathMountInfo string) (bool, error) {
+	var isV2 bool
+	newMountPoint := func(mp *MountPoint) error {
+		if mp.FSType == _cgroupv2FSType && mp.MountPoint == _cgroupv2MountPoint {
+			isV2 = true
+		}
+		return nil
```

Comment:
My knowledge of cgroups is questionable but reading this, I would guess that we shouldn't hard-code the mount point and instead, if the type was "cgroups2", record the mount point and use that path when determining the quota. Is that not desirable?

Reply:
The same thing has been done on runc: https://github.com/opencontainers/runc/blob/main/libcontainer/cgroups/utils.go#L34-L50

If I'm not mistaken, cgroup v2 & v1 can be mounted simultaneously, and in that case, cgroup v2 is mounted on /sys/fs/cgroup/unified. From seeing this code I thought it'd be enough but I can't say for sure.

Reply:
Looks like the linked code differentiates between /sys/fs/cgroup and /sys/fs/cgroup/unified.
Looking around a system running in hybrid mode ([per this stack answer](https://unix.stackexchange.com/a/480748)), it seems to me that if we follow my suggestion, we'll need a second check for whether it's in hybrid mode.
I'm okay with hard-coding the path in that case.